### PR TITLE
Animation: Refactoring to make it easier to use and better separate c…

### PR DESCRIPTION
**Pull request by ResearchDaniel**
…oncerns.

### CODE REVIEW

File: modules/animation/src/datastructures/animation.cpp [hunk 1]
```diff
@@ -144,7 +137,6 @@ BasePropertyTrack* Animation::add(Property* property) {
                     LogWarn(e.getMessage() << " Invalid property class identified?") return nullptr;
                 }
                 add(std::move(track));
-                property->getOwner()->addObserver(this);
```
Comment (petersteneteg) [inline modules/animation/src/datastructures/animation.cpp]: how was this solved instead?

File: modules/animationqt/src/sequenceeditor/propertysequenceeditor.cpp [hunk 2]
```diff
@@ -193,17 +190,16 @@ PropertySequenceEditor::PropertySequenceEditor(KeyframeSequence& sequence, Track
...
     sublayout->addWidget(interpolation_, 0, 1);
 
-    auto map = manager.getInterpolationMapping();
-
-    for (auto range = map.equal_range(bpt.getProperty()->getClassIdentifier());
-         range.first != range.second; ++range.first) {
-        auto ip = manager.getInterpolationFactory().create(range.first->second);
+    auto& valseq = dynamic_cast<ValueKeyframeSequence&>(sequence);
+    for (auto interpObj : valseq.getSupportedInterpolations(manager.getInterpolationFactory())) {
+        auto ip = interpObj->create();
```
Comment (petersteneteg) [inline modules/animationqt/src/sequenceeditor/propertysequenceeditor.cpp:195]: at some point the name should be in the factory object, so we don't have to create them just to find the name...

File: modules/animation/src/datastructures/animation.cpp [hunk 3]
```diff
@@ -71,13 +73,95 @@ void Animation::add(std::unique_ptr<Track> track) {
...
+        propertyTrack->getProperty()->getOwner()->addObserver(this);
+    }
     notifyTrackAdded(tracks_.back().get());
 }
 
+Keyframe* Animation::addKeyframe(Property* property, Seconds time) {
+    auto it = findTrack(property);
+    try {
+        if (it != end()) {
+            // Note: interpolation will only be used if a new sequence is created.
```
Comment (petersteneteg) [inline modules/animation/src/datastructures/animation.cpp]: remove comment

File: modules/animation/src/datastructures/animation.cpp [hunk 4]
```diff
@@ -71,13 +73,95 @@ void Animation::add(std::unique_ptr<Track> track) {
...
+
+BasePropertyTrack* Animation::add(Property* property) {
+    auto it = findTrack(property);
+    if (it != end()) {
+        return dynamic_cast<BasePropertyTrack*>(&(*it));
+    } else {
+        if (auto track = getManager()->getTrackFactory().create(property)) {
+            if (auto basePropertyTrack = dynamic_cast<BasePropertyTrack*>(track.get())) {
+                try {
+                    basePropertyTrack->setProperty(const_cast<Property*>(property));
```
Comment (petersteneteg) [inline modules/animation/src/datastructures/animation.cpp]: the const cast should not be needed

File: modules/animation/src/datastructures/animation.cpp [hunk 5]
```diff
@@ -71,13 +73,95 @@ void Animation::add(std::unique_ptr<Track> track) {
...
+        }
+    });
+}
+
+BasePropertyTrack* Animation::add(Property* property) {
+    auto it = findTrack(property);
+    if (it != end()) {
+        return dynamic_cast<BasePropertyTrack*>(&(*it));
+    } else {
+        if (auto track = getManager()->getTrackFactory().create(property)) {
```
Comment (petersteneteg) [inline modules/animation/src/datastructures/animation.cpp:133]: should we not start observing the owner? or where does that happen?

**Review by petersteneteg — CHANGES_REQUESTED** [review]

File: modules/animation/src/datastructures/animation.cpp (hunk 1 above)
Comment (ResearchDaniel) [reply modules/animation/src/datastructures/animation.cpp]: In add(std::move(track));

File: modules/animationqt/src/sequenceeditor/propertysequenceeditor.cpp (hunk 2 above)
Comment (ResearchDaniel) [reply modules/animationqt/src/sequenceeditor/propertysequenceeditor.cpp:195]: Agree, but not a priority

File: modules/animation/src/datastructures/animation.cpp (hunk 5 above)
Comment (ResearchDaniel) [reply modules/animation/src/datastructures/animation.cpp:133]: We do when the track has been added, `add(std::unique_ptr<Track>)`

File: modules/animation/src/datastructures/animation.cpp (hunk 3 above)
Comment (ResearchDaniel) [reply modules/animation/src/datastructures/animation.cpp]: Fixed

File: modules/animation/src/datastructures/animation.cpp (hunk 1 above)
Comment (petersteneteg) [reply modules/animation/src/datastructures/animation.cpp]: ahh

**Review by petersteneteg — DISMISSED** [review]

**Review by petersteneteg — APPROVED** [review]

**Pull request merged**